subsys/net/lib/tftp/tftp_client.c

```diff
@@ -0,0 +1,524 @@
+			send_err(sock, 0x3, "Buffer Overflow");
+
+			/* Log the error. */
+			LOG_ERR("For the transfer of this file, the user provided \
+				  buffer of size: %d", client->user_buf_size);
+			LOG_ERR("However, it seems that the actual file has larger \
+				  size. Exiting !");
+
+			return TFTPC_BUFFER_OVERFLOW;
+		}
+
+		/* We recevied a "block" of data. Lets update the global
+		 * book-keeping variable that tracks the number of blocks
+		 * received.
+		 */
+		tftpc_block_no++;
+
+		/* Because we are RFC1350 compliant, the block size will
+		 * always be 512. If it is equal to 512, we will assume
+		 * that the transfer is still in progress. If we have
+		 * block size less than 512, we will conclude the transfer.
+		 */
+		return (tftpc_request_size - TFTP_HEADER_SIZE ==
+				TFTP_BLOCK_SIZE) ?
+				(TFTPC_DATA_RX_SUCCESS) :
+				(TFTPC_OP_COMPLETED);
+	} else if (tftpc_block_no > block_no) {
+
+		LOG_INF("Server send duplicate (already acked) block again. \
+				 Block Number: %d", block_no);
+		LOG_INF("Client already trying to receive Block Number: \
+				 %d", tftpc_block_no);
+
+		/* Re-ACK the data. */
+		send_ack(sock, block_no);
+
+		/* OK - This means that we just received a block of data
+		 * that has already been received and acked by the TFTP
+		 * client.
+		 */
+		return TFTPC_DUPLICATE_DATA;
+	}
+
+	/* Don't know what's going on. */
+	return TFTPC_UNKNOWN_FAILURE;
+}
+
+/* Name: tftp_send_request
+ * Description: This function sends out a request to the TFTP Server
+ * (Read / Write) and waits for a response. Once we get some response
+ * from the server, it is interpreted and ensured to be correct.
+ * If not, we keep on poking the server for data until we eventually
+ * give up.
+ */
+static int tftp_send_request(int sock, u8_t request,
+		const char *remote_file, const char *mode)
+{
+	u8_t    no_of_retransmists = 0;
+	s32_t   stat;
+	u16_t   server_response    = -1;
+
+	/* Socket connection successfully - Create the Read Request
+	 * Packet (RRQ).
+	 */
+	make_request(remote_file, mode, request);
+
+	do {
+
+		/* Send this request to the TFTP Server. */
+		stat = send(sock, tftpc_request_buffer, tftpc_request_size, 0);
+
+		/* Data sent successfully? */
+		if (stat > 0) {
+
+			/* We have been able to send the Read Request to the
+			 * TFTP Server. The server will respond to our
+			 * message with data or error, And we need to handle
+			 * these cases correctly.
+			 * Lets try to get this data from the TFTP Server.
+			 */
+			stat = tftpc_recv_data(sock);
+
+			/* Lets check if we were able to get response from the
+			 * server and if the response was Ok or not.
+			 */
+			if (stat > 0) {
+
+				/* Ok - We were able to get response of our read
+				 * request from the TFTP Server. Lets check and
+				 * see what the TFTP Server has to say about our
+				 * request.
+				 */
+				server_response =
+						extract_u16(tftpc_request_buffer);
+
+				/* Did we get some err? */
+				if (server_response == ERROR_OPCODE) {
```

Comment:
I guess refactoring this whole loop content would be nicer, too many code block {} imbrication. We are supposed to follow linux code style, but I have seen patches being applied lately which were not following it (80 chars limit etc...) so what to trust or follow these days...